### PR TITLE
Have AH compute tags function return by reference

### PR DESCRIPTION
**Pull request by carmaza**
## Proposed changes

~I have a couple of extra commits that work out the compute tags in `Tags.hpp` too (in progress). However, I wanted to make sure we actually want to have them as such, given that they are only used as arguments to initialize the prolonged coords in  `ApparentHorizon` (currently in `NumericalAlgorithms/Interpolation/InterpolationTargetApparentHorizon.hpp`). Is the intended idea to give those items their own allocation throughout an evolution at all?~

~Opening this PR in order to see what changes I should actually push.~

Long story short: `NumericalAlgorithms/Interpolation/InterpolationTargetApparentHorizon.hpp` has some allocations used to compute `prolonged_coords` (see `Have compute tags in AH Tags.hpp return by reference` commit). which I'm not sure if can (or should) be avoided.

### Types of changes:

- [ ] Bugfix
- [x ] New feature
- [ ] Refactor

### Component:

- [x ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/ApparentHorizons/ComputeItems.hpp [hunk 1]
```diff
@@ -33,23 +35,29 @@ namespace Tags {
...
+  static void function(
+      const gsl::not_null<tnsr::ii<DataVector, Dim, Frame>*> result,
       const tnsr::aa<DataVector, Dim, Frame>& psi,
       const tnsr::aa<DataVector, Dim, Frame>& pi,
       const tnsr::iaa<DataVector, Dim, Frame>& phi,
       const tnsr::II<DataVector, Dim, Frame>& inv_g) noexcept {
     const auto shift = gr::shift(psi, inv_g);
-    return GeneralizedHarmonic::extrinsic_curvature(
+    destructive_resize_components(result, psi.begin()->size());
+    *result = GeneralizedHarmonic::extrinsic_curvature(
```
Comment (kidder) [inline src/ApparentHorizons/ComputeItems.hpp]: you should be able to call the mutating version of `GeneralizedHarmonic::extrinsic_curvature`

**Review by markscheel — DISMISSED** [review]
Looks good to me.

File: src/ApparentHorizons/ComputeItems.hpp [hunk 2]
```diff
@@ -62,10 +70,13 @@ template <size_t Dim, typename Frame>
...
       db::ComputeTag {
-  static tnsr::Ijj<DataVector, Dim, Frame> function(
+  using return_type = tnsr::Ijj<DataVector, Dim, Frame>;
+  static void function(
+      const gsl::not_null<tnsr::Ijj<DataVector, Dim, Frame>*> result,
       const tnsr::iaa<DataVector, Dim, Frame>& phi,
       const tnsr::II<DataVector, Dim, Frame>& inv_g) noexcept {
-    return raise_or_lower_first_index(
+    destructive_resize_components(result, phi.begin()->size());
+    *result = raise_or_lower_first_index(
```
Comment (kidder) [inline src/ApparentHorizons/ComputeItems.hpp]: should be able to call mutating version of `raise_or_lower_first_index`

**Comment by markscheel** [comment]
You asked about prolonged_coords.  Currently you are doing no more allocations than previously. An alternative (probably left to a future PR if we want to do it) would be to have InterpolationTargetApparentHorizon store a Variables containing all the intermediate results, and then there would be a single allocation at initialization of the InterpolationTarget ParallelComponent.

**Comment by markscheel** [comment]
Actually, yet another alternative would be to allocate a Variables (containing all the intermediate results) just before computing prolonged_coords.  Then there would be no persistent storage, and there would be a single allocation instead of several.

@nilsdeppe opinions on these possibilities for reducing allocations?

**Comment by nilsdeppe** [comment]
@markscheel using a `Variables` is exactly what I was thinking while looking through this PR :) I also agree that it doesn't need to be done as part of this PR

File: src/ApparentHorizons/Tags.hpp [hunk 3]
```diff
@@ -103,10 +110,12 @@ struct InvHessian : db::ComputeTag {
 template <typename Frame>
 struct Radius : db::ComputeTag {
   static std::string name() noexcept { return "Radius"; }
-  SPECTRE_ALWAYS_INLINE static auto function(
-      const ::Strahlkorper<Frame>& strahlkorper) noexcept {
-    return strahlkorper.ylm_spherepack().spec_to_phys(
-        strahlkorper.coefficients());
+  using return_type = DataVector;
+  static void function(const gsl::not_null<DataVector*> radius,
```
Comment (kidder) [inline src/ApparentHorizons/Tags.hpp:113]: let's just move the definition of this to the cpp file to avoid including ContainerHelpers.hpp

File: src/ApparentHorizons/StrahlkorperGr.cpp [hunk 4]
```diff
@@ -449,62 +499,81 @@ tnsr::ii<DataVector, 3, Frame> extrinsic_curvature(
...
+template <typename Frame>
+void ricci_scalar(
+    const gsl::not_null<Scalar<DataVector>*> result,
     const tnsr::ii<DataVector, 3, Frame>& spatial_ricci_tensor,
     const tnsr::I<DataVector, 3, Frame>& unit_normal_vector,
     const tnsr::ii<DataVector, 3, Frame>& extrinsic_curvature,
     const tnsr::II<DataVector, 3, Frame>& upper_spatial_metric) noexcept {
-  auto ricci_scalar = trace(spatial_ricci_tensor, upper_spatial_metric);
+  destructive_resize_components(result, spatial_ricci_tensor.begin()->size());
+  *result = trace(spatial_ricci_tensor, upper_spatial_metric);
```
Comment (kidder) [inline src/ApparentHorizons/StrahlkorperGr.cpp]: is there a mutating version of `trace`?

File: src/ApparentHorizons/StrahlkorperGr.cpp [hunk 5]
```diff
@@ -549,11 +632,22 @@ Scalar<DataVector> euclidean_area_element(
     }
   }
 
-  auto area_element = Scalar<DataVector>{square(radius)};
-  get(area_element) *= sqrt(get(dot_product(cap_theta, cap_theta)) *
-                                get(dot_product(cap_phi, cap_phi)) -
-                            square(get(dot_product(cap_theta, cap_phi))));
-  return area_element;
+  *result = Scalar<DataVector>{square(radius)};
```
Comment (kidder) [inline src/ApparentHorizons/StrahlkorperGr.cpp]: `get(*result) = square(radius);`

File: src/ApparentHorizons/StrahlkorperGr.cpp [hunk 6]
```diff
@@ -520,20 +589,34 @@ Scalar<DataVector> area_element(
     }
   }
 
-  auto area_element = Scalar<DataVector>{square(radius)};
-  get(area_element) *=
+  *result = Scalar<DataVector>{square(radius)};
```
Comment (kidder) [inline src/ApparentHorizons/StrahlkorperGr.cpp]: get(*result) = square(radius);

File: src/ApparentHorizons/StrahlkorperGr.cpp [hunk 7]
```diff
@@ -360,56 +361,93 @@ double get_spin_magnitude(const std::array<DataVector, 3>& potentials,
 namespace StrahlkorperGr {
 
 template <typename Frame>
-tnsr::i<DataVector, 3, Frame> unit_normal_one_form(
+void unit_normal_one_form(
+    const gsl::not_null<tnsr::i<DataVector, 3, Frame>*> result,
     const tnsr::i<DataVector, 3, Frame>& normal_one_form,
     const DataVector& one_over_one_form_magnitude) noexcept {
-  auto unit_normal_one_form = normal_one_form;
+  destructive_resize_components(result, normal_one_form.begin()->size());
```
Comment (kidder) [inline src/ApparentHorizons/StrahlkorperGr.cpp]: should not need to do the destructive_resize_components when setting `*result =` an existing object as operator= should do any resize it needs, so omit it here

File: src/ApparentHorizons/StrahlkorperGr.cpp [hunk 8]
```diff
@@ -360,56 +361,93 @@ double get_spin_magnitude(const std::array<DataVector, 3>& potentials,
...
+  return result;
+}
+
+template <typename Frame>
+void inverse_surface_metric(
+    const gsl::not_null<tnsr::II<DataVector, 3, Frame>*> result,
     const tnsr::I<DataVector, 3, Frame>& unit_normal_vector,
     const tnsr::II<DataVector, 3, Frame>& upper_spatial_metric) noexcept {
-  auto inv_surf_metric = upper_spatial_metric;
+  destructive_resize_components(result, unit_normal_vector.begin()->size());
```
Comment (kidder) [inline src/ApparentHorizons/StrahlkorperGr.cpp]: don't need

File: src/ApparentHorizons/StrahlkorperGr.cpp [hunk 9]
```diff
@@ -424,23 +462,35 @@ Scalar<DataVector> expansion(
...
 }
 
 template <typename Frame>
-tnsr::ii<DataVector, 3, Frame> extrinsic_curvature(
+void extrinsic_curvature(
+    const gsl::not_null<tnsr::ii<DataVector, 3, Frame>*> result,
     const tnsr::ii<DataVector, 3, Frame>& grad_normal,
     const tnsr::i<DataVector, 3, Frame>& unit_normal_one_form,
     const tnsr::I<DataVector, 3, Frame>& unit_normal_vector) noexcept {
+  destructive_resize_components(result, grad_normal.begin()->size());
```
Comment (kidder) [inline src/ApparentHorizons/StrahlkorperGr.cpp]: don't need

File: src/ApparentHorizons/StrahlkorperGr.cpp [hunk 10]
```diff
@@ -449,62 +499,81 @@ tnsr::ii<DataVector, 3, Frame> extrinsic_curvature(
...
+
+template <typename Frame>
+void ricci_scalar(
+    const gsl::not_null<Scalar<DataVector>*> result,
     const tnsr::ii<DataVector, 3, Frame>& spatial_ricci_tensor,
     const tnsr::I<DataVector, 3, Frame>& unit_normal_vector,
     const tnsr::ii<DataVector, 3, Frame>& extrinsic_curvature,
     const tnsr::II<DataVector, 3, Frame>& upper_spatial_metric) noexcept {
-  auto ricci_scalar = trace(spatial_ricci_tensor, upper_spatial_metric);
+  destructive_resize_components(result, spatial_ricci_tensor.begin()->size());
```
Comment (kidder) [inline src/ApparentHorizons/StrahlkorperGr.cpp]: shouldn't need if mutating version calls `destructive_resize_components` or if there is not a mutating version

File: src/ApparentHorizons/StrahlkorperGr.cpp [hunk 11]
```diff
@@ -520,20 +589,34 @@ Scalar<DataVector> area_element(
...
+}
+
+template <typename Frame>
+void euclidean_area_element(
+    const gsl::not_null<Scalar<DataVector>*> result,
     const StrahlkorperTags::aliases::Jacobian<Frame>& jacobian,
     const tnsr::i<DataVector, 3, Frame>& normal_one_form,
     const DataVector& radius,
     const tnsr::i<DataVector, 3, Frame>& r_hat) noexcept {
+  destructive_resize_components(result, radius.size());
```
Comment (kidder) [inline src/ApparentHorizons/StrahlkorperGr.cpp]: don't need

File: src/ApparentHorizons/StrahlkorperGr.cpp [hunk 12]
```diff
@@ -449,62 +499,81 @@ tnsr::ii<DataVector, 3, Frame> extrinsic_curvature(
...
-    const tnsr::i<DataVector, 3, Frame>& normal_one_form,
-    const DataVector& radius,
-    const tnsr::i<DataVector, 3, Frame>& r_hat) noexcept {
+void area_element(const gsl::not_null<Scalar<DataVector>*> result,
+                  const tnsr::ii<DataVector, 3, Frame>& spatial_metric,
+                  const StrahlkorperTags::aliases::Jacobian<Frame>& jacobian,
+                  const tnsr::i<DataVector, 3, Frame>& normal_one_form,
+                  const DataVector& radius,
+                  const tnsr::i<DataVector, 3, Frame>& r_hat) noexcept {
+  destructive_resize_components(result, radius.size());
```
Comment (kidder) [inline src/ApparentHorizons/StrahlkorperGr.cpp]: don't need

File: src/ApparentHorizons/StrahlkorperGr.cpp [hunk 13]
```diff
@@ -577,19 +671,24 @@ double euclidean_surface_integral_of_vector(
...
+    component = 0.0;
+  }
+
   auto extrinsic_curvature_theta_normal_sin_theta =
       make_with_value<Scalar<DataVector>>(area_element, 0.0);
   auto extrinsic_curvature_phi_normal =
       make_with_value<Scalar<DataVector>>(area_element, 0.0);
 
-  DataVector& extrinsic_curvature_dot_normal = get(temp);
+  DataVector& extrinsic_curvature_dot_normal = get(*result);
```
Comment (kidder) [inline src/ApparentHorizons/StrahlkorperGr.cpp:683]: add comment // using result as temporary

File: src/ApparentHorizons/StrahlkorperGr.cpp [hunk 14]
```diff
@@ -613,18 +712,28 @@ Scalar<DataVector> spin_function(
         extrinsic_curvature_dot_normal * tangents.get(i, 1);
   }
 
-  DataVector& sin_theta = get(temp);
+  DataVector& sin_theta = get(*result);
```
Comment (kidder) [inline src/ApparentHorizons/StrahlkorperGr.cpp:708]: add comment // using result as temporary

File: src/ApparentHorizons/StrahlkorperGr.cpp [hunk 15]
```diff
@@ -613,18 +712,28 @@ Scalar<DataVector> spin_function(
...
   sin_theta = sin(ylm.theta_phi_points()[0]);
   get(extrinsic_curvature_theta_normal_sin_theta) *= sin_theta;
   get(extrinsic_curvature_phi_normal) *= sin_theta;
 
-  Scalar<DataVector>& spin_function = temp;
-  get(spin_function) =
+  get(*result) =
```
Comment (kidder) [inline src/ApparentHorizons/StrahlkorperGr.cpp:714]: add comment // now computing actual result

**Review by kidder — CHANGES_REQUESTED** [review]
Thank you for doing this!  reviewed with @nilsdeppe

File: src/NumericalAlgorithms/Interpolation/InterpolationTargetApparentHorizon.hpp [hunk 16]
```diff
@@ -202,9 +202,14 @@ struct ApparentHorizon {
...
-    StrahlkorperTags::aliases::OneForm<Frame> r_hat{};
-    DataVector radius{};
+    Variables<tmpl::list<::Tags::Tempi<0, 2, ::Frame::Spherical<Frame>>,
+                         ::Tags::Tempi<1, 3, Frame>, ::Tags::TempScalar<2>>>
+        temp_buffer(prolonged_strahlkorper.ylm_spherepack().physical_size());
+
+    auto& theta_phi =
+        get<::Tags::Tempi<0, 2, ::Frame::Spherical<Frame>>>(temp_buffer);
+    auto& r_hat = get<::Tags::Tempi<1, 3, Frame>>(temp_buffer);
+    auto& radius = get(get<::Tags::TempScalar<2>>(temp_buffer));
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/InterpolationTargetApparentHorizon.hpp:212]: Thanks for doing this ❤️

**Review by kidder — COMMENTED** [review]
you can squash

**Comment by carmaza** [comment]
Squashed and rebased on develop.

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]
image

Thank you so much doing this :) Again, very much appreciated :D
